### ISSUE TITLE
[HOLD for payment 2025-02-04] [$250] Search - Expense from the list does not disappear after deletion

### ISSUE THREAD

**Issue by IuliiaHerets** [Expensify/App#53687]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.72-0
**Reproducible in staging?:** Yes
**Reproducible in production?:** Yes
**If this was caught on HybridApp, is this reproducible on New Expensify Standalone?:** Yes, reproducible on both
**If this was caught during regression testing, add the test name, ID and link from TestRail:** https://expensify.testrail.io/index.php?/tests/view/5309274
**Email or phone of affected tester (no customers):** sustinov@applausemail.com
**Issue reported by:** Applause Internal Team

## Action Performed:
1. Open HybridApp
2. Log in to the HT account
3. Send 2 manual expenses to any user
4. Go to Search -> Expenses -> All
5. Delete one of the created expenses from step 3
6. Tap again on the deleted expense

## Expected Result:
After deleting an expense, it should immediately disappear from the list without refreshing the page

## Actual Result:
Expense from the list does not disappear after deletion

## Workaround:
Unknown

## Platforms:
- [x] Android: Standalone
- [x] Android: HybridApp
- [x] Android: mWeb Chrome
- [x] iOS: Standalone
- [x] iOS: HybridApp
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/917c7a77-9d7d-4aaf-8cfa-25d030038739

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021865183017070238528</li>
        <li>Upwork Job ID: 1865183017070238528</li>
        <li>Last Price Increase: 2024-12-13</li>
<li>Automatic offers: </li>
<ul>
<li>rayane-djouah | Reviewer | 105373332</li>
<li>FitseTLT | Contributor | 105373334</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @mallenexpensify</details>

**Comment by FitseTLT** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Search - Expense from the list does not disappear after deletion

### What is the root cause of that problem?
We only send search request if there is an increase in transaction length, i.e. the currentTransaction length is greater than the previous transaction length
https://github.com/Expensify/App/blob/b9107bf59cbb3982d958abbae357e3b903ed4c63/src/hooks/useSearchHighlightAndScroll.ts#L42
but here we are deleting the  expense that will decrease the transactions length we have 
### What changes do you think we should make in order to solve the problem?

The conditions that depend on transaction length are incorrect in both
https://github.com/Expensify/App/blob/b9107bf59cbb3982d958abbae357e3b903ed4c63/src/hooks/useSearchHighlightAndScroll.ts#L42
https://github.com/Expensify/App/blob/b9107bf59cbb3982d958abbae357e3b903ed4c63/src/hooks/useSearchHighlightAndScroll.ts#L37
because transaction length equality can even mean there is a change that needs a search request to be sent. For instance, if a user creates a new expense but deletes another expense which could result in the transaction lengths to be equal but still there is a change that requires a fetch.
We should instead depend on a condition where the previous and current transaction IDs list are not equal (of course with deep comparison between the values they hold)
https://github.com/Expensify/App/blob/b9107bf59cbb3982d958abbae357e3b903ed4c63/src/hooks/useSearchHighlightAndScroll.ts#L76-L80
```
const previousTransactionIDs = previousTransactions && Object.keys(previousTransactions);
        const transactionIDs = transactions && Object.keys(transactions);
```
and replace this with it
https://github.com/Expensify/App/blob/b9107bf59cbb3982d958abbae357e3b903ed4c63/src/hooks/useSearchHighlightAndScroll.ts#L42
```
if(!isEqual(transactionIDs, previousTransactionIDs))
```
and remove the condition here
https://github.com/Expensify/App/blob/b9107bf59cbb3982d958abbae357e3b903ed4c63/src/hooks/useSearchHighlightAndScroll.ts#L37
```
        if (searchTriggeredRef.current) {

```

### What specific scenarios should we cover in automated tests to prevent reintroducing this issue in the future?
We can render the Search page (or optionally test only `useSearchHighlightAndScroll`) and change the transaction list from onyx by replacing a transaction with another one and mock our `SearchActions.search` and assert that it has been called 👍 
### What alternative solutions did you explore? (Optional)

**Comment by mallenexpensify** [comment]
@rayane-djouah can you attempt reproduction and, if you're able to, review the above proposals? I tried to test and staging.new.expensify.com froze for me :/

**Comment by luacmartins** [comment]
I agree. Thanks for the detailed summary!

**Comment by luacmartins** [comment]
Ok, I have PRs to update the responses from `DeleteMoneyRequest` and `DeleteMoneyRequestOnSearch`. Below is the sample response:

When deleting the last transaction on a report (no snapshot key is sent, since live updates should automatically update the snapshot)
![Screenshot 2025-01-10 at 12 56 08 PM](https://github.com/user-attachments/assets/8331c7e4-9ae6-4c4c-9f28-2c22a39df29c)

When deleting a single transaction on a report with multiple expenses (snapshot key is returned to update isOneTransactionReport, etc)

![Screenshot 2025-01-10 at 12 55 47 PM](https://github.com/user-attachments/assets/b965d867-8256-49c7-a4e3-6ae1e8bcfd27)

@rayane-djouah @FitseTLT Does that update look correct to you?

**Comment by FitseTLT** [comment]
@luacmartins I am a bit confused here. Currently useSearchHighlightAndScroll hook checks for a change in `transactions_` key and send search request accordingly. Now the problem we are facing is `transactions_` key is not being correctly updated on deleting the last transaction in a report I was hoping to get that `transactions_` key update fixed on this case so that the normal operation of the hook will work for this case too.
I am not sure how the BE code works but I don't think fixing the response of DeleteRequest API only suffice but also the data it pushes too for other users that didn't delete the money request. Like User A deleted a request in workspace that is owned by UserB and now the update doesn't reach for both users.

**Comment by FitseTLT** [comment]
> @FitseTLT yes, the data is being pushed to other users, but for some reason we're not returning it in the http response to the caller. So my PRs are adding all the onyxData updates to the http response. Does that make sense?

Yep. But if you want to test you can do with branch of the [PR](https://github.com/Expensify/App/pull/54407) because the BE change should solve it without any more FE change :+1:

**Comment by luacmartins** [comment]
@FitseTLT doesn't seem to work. Check out the video below:

https://github.com/user-attachments/assets/19d19cd5-4656-44a9-ad99-5896b7b094d2

**Comment by FitseTLT** [comment]
@luacmartins Did you apply any BE changes? Because now it is working I remember invoice deletion was not updated too but it is working now.

https://github.com/user-attachments/assets/ccaf49c5-4adc-4927-8877-f7cc5d283c82


https://github.com/user-attachments/assets/c45fde0e-6ad6-48fa-9bac-b91c301a260d

**Comment by luacmartins** [comment]
No BE changes that I'm aware of.

**Comment by FitseTLT** [comment]
> No BE changes that I'm aware of.

Ok but now the BE is correctly updating transactions_ list so it seems fixed. I will work on the rest of the tasks left but let u know if any BE change is needed. Maybe some other BE PR might have fixed it Thx 👍

**Comment by mallenexpensify** [comment]
Contributor: @FitseTLT paid $250 via Upwork
Contributor+: @rayane-d paid $250 via Upwork 

Test case - https://github.com/Expensify/Expensify/issues/467599

Thx!
